client/orderbook/bookside.go

```diff
@@ -142,7 +152,9 @@ func (d *bookSide) BestNOrders(n int) ([]*Order, bool) {
 		if count == 0 {
 			return false
 		}
-		best = append(best, ord)
+		// Return copies for thread-safe access to the Quantity field.
+		ordCopy := *ord
+		best = append(best, &ordCopy)
```

Comment:
I felt this was preferable to making multiple copies of each Order, and doing it those copies on each add/book.

Reply:
We were only creating copies when the `Remaining` field was changed, which isn't going to happen for every order. Or maybe I'm not following what you mean by "making multiple copies... and doing those copies on each add/book".

The proposed change guarantees that every order is copied every time it is requested, which could be repeatedly if the user e.g. is switching back and forth between a couple of markets.

Reply:
Mmm, you're right, sorry about that bad assertion... it's the update that creates copies, not add/book.  What happens more frequently though, updating orders or requesting the fresh book?  Subscriptions and the associated orderbooks stick around for a bit before being deleted I believe.

Regardless, I'm not married to copying one place or the other, but part of what bothered me is that there was not previously a need to update the orders map at all on updateRemaining since the the purpose of the map appears to be: index of all orders in the book sides, and remember each one's rate for bookSide.Remove and side... while the bookSide deals with remaining exclusively. But copy-on-update requires updating the orders map unless we let those entries just have invalid remaining amounts.  Also I felt like UpdateRemaining vs ReplaceOrder was clearer about the purpose of the action.

I'll revisit this again, going back to `ReplaceOrder`, just with an actual copy (and the other bug fixes), but this `orders` map fundamentally bothers me for some reason.

Reply:
> this `orders` map fundamentally bothers me for some reason.

Probably because its purpose is to avoid having to iterate the `bookSide.bins`. And that would be great if we weren't passing the `*Order` out directly. So either we create/copy the `*Order`s on every call to `(*OrderBook).{Orders, BestFill, ...}`, or we iterate `bookSide.bins` to replace the order during `update_remining`. I do think the second option is more efficient, but I'm not digging my heels in on this one, so if you're hardcore to copy, go for it. 

We could almost get away with just indexing the order id -> rate rather than the whole `*Order`, but we are using the `Side` field in [one other place](https://github.com/chappjc/dcrdex/blob/be02613c3ce6bdc09a81a2b0b78e4334f290b8b6/client/orderbook/orderbook.go#L415-L426). Could probably be worked around though.

Reply:
> We could almost get away with just indexing the order id -> rate rather than the whole `*Order`, but we are using the `Side` field in [one other place](https://github.com/chappjc/dcrdex/blob/be02613c3ce6bdc09a81a2b0b78e4334f290b8b6/client/orderbook/orderbook.go#L415-L426). Could probably be worked around though

This had occurred to me too.  That is making `orders map[OrderID]struct{rate uint64; sell bool}`.  I'm going to get to this PR today.

Reply:
Back to making a copy on `update_remaining`, but it's done inside `bookSide` now. The `OrderBook.orders` map no longer stores the `*Order`, which was my main hangup with the map.  It indexes the orders and keeps the info needed to locate the order in a `bookSide` (rate and side), nothing more.

Reply:
Oh silly me I forgot to remove this copy.